File: vtem_control_matlab/VtemControl.m
classdef VtemControl < handle
   properties
      DeviceAddress_
      Port_
      ctx_
      connected_ = false;
      addr_input_start_ = 45392;
      addr_output_start_ = 40001;
      cpx_input_offset_ = 4;
      cpx_output_offset_ = 3;
      num_slots_ = 8;
   end
   methods(Static)
      function [slotIdx, slotRemain] = get_slot_idx_from_valve_idx(valveIdx)
         slotIdx = floor(valveIdx / 2);
         slotRemain = valveIdx - 2*slotIdx;
      end
   end
   methods
      function obj = VtemControl(DeviceAddress, Port)
         obj.DeviceAddress_ = DeviceAddress;
         obj.Port_ = Port;
      end
      function connect(obj)
         obj.ctx_ = modbus('tcpip', obj.DeviceAddress_, obj.Port_);
         obj.connected_ = true;
      end
      function disconnect(obj)
         % clear obj.ctx_;
         delete(obj.ctx_);
         obj.connected_ = false;
      end
      function ensure_connection(obj)
         if obj.connected_ ~= true
             throw(MException("VtemControl:ensure_connection", "Operation requires a modbus connection."))
         end
      end
      function [value, bits_str] = read_address(obj, addr)
         value = read(obj.ctx_, 'holdingregs', addr, 1, 'int16');
         bits = bitget(value, 8:-1:1);
         bits_str = num2str(bits);
         bits_str(isspace(bits_str)) = '';
      end
      function [motion_app_id, valve_state] = get_single_motion_app(obj, slotIdx)
         obj.ensure_connection();
          
         % we read two bytes for all 3 entries 
         % (e.g. status, actual value 1, actual value 2)
         addr = obj.addr_input_start_ + obj.cpx_input_offset_ + 3*slotIdx;
         % this gives us an array containing two bytes
         status = read(obj.ctx_, 'holdingregs', addr, 1, 'int16');
         
%          addr1 = obj.addr_input_start_ + obj.cpx_input_offset_ + 2*3*slotIdx
%          byte1 = read(obj.ctx_, 'holdingregs', addr1, 1);
%          byte1_bits = bitget(byte1, 8:-1:1, 'int16')
%          addr2 = obj.addr_input_start_ + obj.cpx_input_offset_ + 2*3*slotIdx + 1
%          byte2 = read(obj.ctx_, 'holdingregs', addr2, 1);
%          byte2_bits = bitget(byte2, 8:-1:1, 'int16')
         
%          byte1_bits = bitget(status(2), 8:-1:1, 'int16')
%          byte2_bits = bitget(status(1), 8:-1:1, 'int16')
         
         % somehow, the second byte seems to be stored in the first element
         % of the array and vice-versa
         motion_app_id_bits = bitget(status, 6:-1:1, 'int16');
         motion_app_id = bit2dec(motion_app_id_bits);
         
         valve_state_bits = bitget(status, 8:-1:7, 'int16');
         valve_state = bit2dec(valve_state_bits);
         
         app_state_bits = bitget(status, 16:-1:9, 'int16');
         app_state = bit2dec(app_state_bits);
      end
      function set_single_motion_app(obj, slotIdx, motion_app_id, app_control)
         obj.ensure_connection();
          
         app_option = 0;
         
         motion_app_id_bin = bitget(motion_app_id, 6:-1:1, 'int16');
         app_control_bin = bitget(app_control, 2:-1:1, 'int16');
         app_option_bin = bitget(app_option, 8:-1:1, 'int16');
         
         command_bits = [app_option_bin app_control_bin motion_app_id_bin];
         command = bit2dec(command_bits);
         
         addr = obj.addr_output_start_ + obj.cpx_output_offset_ + 3*slotIdx;
         write(obj.ctx_, 'holdingregs', addr, command, 'int16');
      end
      function set_all_motion_apps(obj, motion_app_id, app_control)
          for slotIdx = 0:1:(obj.num_slots_-1)
              obj.set_single_motion_app(slotIdx, motion_app_id, app_control);
          end
      end
      function result = ensure_motion_app(obj, slotIdx, des_motion_app_id, des_valve_state, throw_exception)
         if nargin < 5
            throw_exception = true;
         end

         [motion_app_id, valve_state] = get_single_motion_app(obj, slotIdx);
          
         if motion_app_id ~= des_motion_app_id
             if throw_exception
                throw(MException("VtemControl:ensure_motion_app", "Operation requires activating the suitable motion app."))
             else
                result = false;
                return;
             end
         end
         
         if valve_state ~= des_valve_state
             if throw_exception
                throw(MException("VtemControl:ensure_motion_app", "Operation requires setting the desired valve state."))
             else
                result = false;
                return;
             end
         end

         result = true;
         return;
      end
      function acknowledge_errors_single_slot(obj, slotIdx)
         obj.set_single_motion_app(slotIdx, 62, 1);
         % Wait for errors to be acknowledged
         pause(0.5);
      end
      function acknowledge_errors_all_slots(obj)
         obj.set_all_motion_apps(62, 1);
         % Wait for errors to be acknowledged
         pause(0.5);
      end
      function result = activate_pressure_regulation_single_slot(obj, slotIdx)
         des_valve_mode = 3;
         des_app_control = 3;
         des_valve_state = 2;
         timeout = 20; % [s] duration during which we wait for the pressure regulation to deactivate
         sampling_rate = 20; % frequency with which we check the currently activate valve mode
         
         obj.set_single_motion_app(slotIdx, des_valve_mode, des_app_control);

         i = 0;
         config_mismatch = true;
         while config_mismatch
            if i / sampling_rate > timeout
                result = false;
                return;
            end

            [actual_valve_mode, valve_state] = obj.get_single_motion_app(slotIdx);
            
            if actual_valve_mode == des_valve_mode && valve_state == des_valve_state
                config_mismatch = false;
            end

            if config_mismatch
                i = i + 1;
                pause(1/sampling_rate);
            end
         end

         result = true;
         return;
      end
      function result = deactivate_pressure_regulation_single_slot(obj, slotIdx)
         des_valve_mode = 61;
         des_app_control = 0;
         des_valve_state = 1;
         timeout = 10; % [s] duration during which we wait for the pressure regulation to deactivate
         sampling_rate = 20; % frequency with which we check the currently activate valve mode
         exhaust_duration = 1; % [s] duration to let valve exhaust before shutting off motion app

         % Set valves to 0 bar (off).
         obj.set_single_pressure(2*slotIdx, 0);
         obj.set_single_pressure(2*slotIdx + 1, 0);

         % take some time to release pressure before shutting of the motion app
         pause(exhaust_duration);
         
         obj.set_single_motion_app(slotIdx, des_valve_mode, des_app_control);

         i = 0;
         config_mismatch = true;
         while config_mismatch
            if i / sampling_rate > timeout
                result = false;
                return;
            end

            [actual_valve_mode, valve_state] = obj.get_single_motion_app(slotIdx);
            
            if actual_valve_mode == des_valve_mode && valve_state == des_valve_state
                config_mismatch = false;
            end

            if config_mismatch
                i = i + 1;
                pause(1/sampling_rate);
            end
         end

         result = true;
         return;
      end
      function result = activate_pressure_regulation_all_slots(obj)
         des_valve_mode = 3;
         des_app_control = 3;
         des_valve_state = 2;
         timeout = 20; % [s] duration during which we wait for the pressure regulation to deactivate
         sampling_rate = 20; % frequency with which we check the currently activate valve mode
         
         obj.set_all_motion_apps(des_valve_mode, des_app_control);

         i = 0;
         config_mismatch = true;
         while config_mismatch
            if i / sampling_rate > timeout
                result = false;
                return;
            end

            config_mismatch = false;
            for slotIdx = 0:1:(obj.num_slots_-1)
                [actual_valve_mode, valve_state] = obj.get_single_motion_app(slotIdx);
                
                if actual_valve_mode ~= des_valve_mode || valve_state ~= des_valve_state
                    config_mismatch = true;
                end
            end

            if config_mismatch
                i = i + 1;
                pause(1 / sampling_rate);
            end
         end

         result = true;
         return;
      end
      function result = deactivate_pressure_regulation_all_slots(obj)
         des_valve_mode = 61;
         des_app_control = 0;
         des_valve_state = 1;
         timeout = 10; % [s] duration during which we wait for the pressure regulation to deactivate
         sampling_rate = 20; % frequency with which we check the currently activate valve mode
         exhaust_duration = 1; % [s] duration to let valve exhaust before shutting off motion app

         % Set valves to 0 bar (off).
         for slotIdx = 0:1:(obj.num_slots_-1)
            obj.set_single_pressure(2*slotIdx, 0);
            obj.set_single_pressure(2*slotIdx + 1, 0);
         end

         % take some time to release pressure before shutting of the motion app
         wait(exhaust_duration);
         
         obj.set_all_motion_apps(des_valve_mode, des_app_control);

         i = 0;
         config_mismatch = true;
         while config_mismatch
            if i / sampling_rate > timeout
                result = false;
                return;
            end

            config_mismatch = false;
            for slotIdx = 0:1:(obj.num_slots_-1)
                [actual_valve_mode, valve_state] = obj.get_single_motion_app(slotIdx);
                
                if actual_valve_mode ~= des_valve_mode || valve_state ~= des_valve_state
                    config_mismatch = true;
                end
            end

            if config_mismatch
                i = i + 1;
                pause(1 / sampling_rate);
            end
         end

         result = true;
         return;
      end
      function value = get_single_pressure(obj, valveIdx)
         obj.ensure_connection(); 
         
         [slotIdx, slotRemain] = VtemControl.get_slot_idx_from_valve_idx(valveIdx);
         
         if obj.ensure_motion_app(slotIdx, 3, 2, false) == false
             % we return zero pressure if the pressure regulation app is not active
            value = 0;
            return;
         end
          
         addr = obj.addr_input_start_ + obj.cpx_input_offset_ + 3*slotIdx + 1 + slotRemain;
         bytes = read(obj.ctx_, 'holdingregs', addr, 1, 'int16');

         bits = bitget(bytes, 16:-1:1, 'int16');
         value = bit2dec(bits);
      end
      function set_single_pressure(obj, valveIdx, value)
         obj.ensure_connection();
         
         [slotIdx, slotRemain] = VtemControl.get_slot_idx_from_valve_idx(valveIdx);
         
         obj.ensure_motion_app(slotIdx, 3, 2);
          
         addr = obj.addr_output_start_ + obj.cpx_output_offset_ + 3*slotIdx + 1 + slotRemain;
         write(obj.ctx_, 'holdingregs', addr, value, 'int16');
      end
      function pressures = get_all_pressures(obj)
          pressures = zeros(2*obj.num_slots_, 1);
          for valveIdx = 0:1:(2*obj.num_slots_-1)
              pressures(valveIdx + 1) = obj.get_single_pressure(valveIdx);
          end
      end
      function set_all_pressures(obj, pressures)
          for valveIdx = 0:1:(2*obj.num_slots_-1)
              obj.set_single_pressure(valveIdx, pressures(valveIdx + 1));
          end
      end
   end
end

File: vtem_control_cpp/src/VtemControl.cpp
#include "modbus/modbus-tcp.h"
#include "VtemControl.hpp"

#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <iostream>
#include <stdexcept>
#include <thread>


namespace {
    const int address_input_start = 45392;
    const int address_output_start = 40001;
    const int cpx_input_offset = 3;
    const int cpx_output_offset = 2;
    const int num_slots = 8;
}

vtem_control::VtemControl::VtemControl(const char *node, const char *service) {
    // Not connected.
    connected_ = false;

    // Resize buffer space.
    // Input buffer for each valve is of the format (actual, setpoint, diagnostic).
    input_status_buffer_.resize(2 * num_slots);
    input_value_buffer_.resize(2 * 2 * num_slots);

    // Create Modbus context.
    ctx_ = modbus_new_tcp_pi(node, service);
}

vtem_control::VtemControl::~VtemControl() {
    // Close if not yet closed.
    if (connected_) {
        disconnect();
    }

    // Deallocate Modbus context.
    modbus_free(ctx_);
}

bool vtem_control::VtemControl::connect() {
    if (modbus_connect(ctx_) == -1) {
        std::cout << "Connection failed" << std::endl;
        connected_ = false;
        return false;
    }
    connected_ = true;
    return true;
}

bool vtem_control::VtemControl::disconnect() {
    if (connected_) {
        modbus_close(ctx_);
        connected_ = false;
        return true;
    }

    return false;
}

bool vtem_control::VtemControl::get_single_motion_app(int slot_idx, int &motion_app_id, int &valve_state) {
    ensure_connection();

    const auto status = &input_status_buffer_[slot_idx]; // this should read two bytes containing the slot status information
    const auto addr = address_input_start + cpx_input_offset + 3*slot_idx;
    
    if (modbus_read_registers(ctx_, addr, 1, (uint16_t *)status) == -1) {
        throw std::runtime_error("Failed to read slot status register.");
    }

    /* example to extract individual bytes: 
    // mask: 0xFF (only selects first byte)
    // shift operator: delete the first 8 bits
    uint8_t first_byte = status & 0xFF;
    uint8_t second_byte = (status >> 8) & 0xFF;
    */

    motion_app_id = *status & 0x3F;
    valve_state = (*status >> 6) & 0x03;
 
    /*  extract bits 6-8 from first byte to read actual valve state
        for motion app 03:
            00: both valves are inactive
            01: second valve is active
            02: first valve is active
            03: both valves are active
    */

    return true;
}

bool vtem_control::VtemControl::set_single_motion_app(int slot_idx, int motion_app_id = 61, int app_control = 0) {
    /*  App control for motion app 03 (proportional pressure regulation):
            00: both valves are inactive
            01: second valve is active
            02: first valve is active
            03: both valves are active
    */

    ensure_connection();

    int16_t command_first_byte = (app_control << 6) | motion_app_id;
    int16_t command_second_byte = 0;
    int16_t command = (command_second_byte << 8) | command_first_byte;
    
    const auto addr = address_output_start + cpx_output_offset + 3*slot_idx;
    if (modbus_write_register(ctx_, addr, command) == -1) {
        throw std::runtime_error("Failed to write register for setting motion app.");
    }

    return true;
}

bool vtem_control::VtemControl::set_all_motion_apps(int motion_app_id = 61, int app_control = 0) {
    for (auto slot_idx = 0; slot_idx < (num_slots); slot_idx++) {
        if (!set_single_motion_app(slot_idx, motion_app_id, app_control)) {
            return false;
        }
    }
    return true;
}

bool vtem_control::VtemControl::acknowledge_errors(int slot_idx = -1) {
    int des_valve_mode = 62;
    int des_app_control = 1;

    if (slot_idx == -1) {
        set_all_motion_apps(des_valve_mode, des_app_control);
    } else {
        set_single_motion_app(slot_idx, des_valve_mode, des_app_control);
    }

    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    return true;
}

bool vtem_control::VtemControl::activate_pressure_regulation(int slot_idx = -1) {
    int des_valve_mode = 3;
    int des_app_control = 3;
    int des_valve_state = 2;
    float timeout = 20; // [s] duration during which we wait for the pressure regulation to deactivate
    float sampling_rate = 100; // frequency with which we check the currently activate valve mode

    if (slot_idx == -1) {
        set_all_motion_apps(des_valve_mode, des_app_control);

        int actual_valve_mode, valve_state;

        int i = 0;
        // wait to reach valve mode 3 (pressure regulation) and valve state 2 (e.g. running)
        bool config_mismatch = true;
        while (config_mismatch) {
            if (i / sampling_rate > timeout) {
                return false;
            }

            config_mismatch = false;
            for (auto slot_idx = 0; slot_idx < (num_slots); slot_idx++) {
                get_single_motion_app(slot_idx, actual_valve_mode, valve_state);

                // std::cout << "Activate pressure reguluation idx: " << i << " for slot: " << slot_idx << " actual_valve_mode: " << actual_valve_mode << " valve_state: " << valve_state << std::endl;
                
                if (actual_valve_mode != des_valve_mode || valve_state != des_valve_state) {
                    config_mismatch = true;
                    break;
                }
            }

            if (config_mismatch) {
                i += 1;
                std::this_thread::sleep_for(std::chrono::milliseconds(int(1/sampling_rate*1000)));
            }
        }
        return true;
    } else {
        set_single_motion_app(slot_idx, des_valve_mode, des_app_control);

        int actual_valve_mode, valve_state;
        get_single_motion_app(slot_idx, actual_valve_mode, valve_state);

        int i = 0;
        // wait to reach valve mode 3 (pressure regulation) and valve state 2 (e.g. running)
        while (actual_valve_mode != des_valve_mode || valve_state != des_valve_state) {
            if (i / sampling_rate > timeout) {
                return false;
            }

            get_single_motion_app(slot_idx, actual_valve_mode, valve_state);

            // std::cout << "Activate pressure reguluation idx: " << i << " actual_valve_mode: " << actual_valve_mode << " valve_state: " << valve_state << std::endl;

            i += 1;
            std::this_thread::sleep_for(std::chrono::milliseconds(int(1/sampling_rate*1000)));
        }
        return true;
    }
}

bool vtem_control::VtemControl::deactivate_pressure_regulation(int slot_idx = -1) {
    int des_valve_mode = 61;
    int des_app_control = 0;
    int des_valve_state = 1;
    float timeout =  10; // [s] duration during which we wait for the pressure regulation to deactivate
    float sampling_rate = 100; // frequency with which we check the currently activate valve mode
    float exhaust_duration = 1; // [s] duration to let valve exhaust before shutting off motion app

    if (slot_idx == -1) {
        // Set valves to 0 bar (off).
        for (auto slot_idx = 0; slot_idx < (num_slots); slot_idx++) {
            set_single_pressure(2*slot_idx, 0);
            set_single_pressure(2*slot_idx + 1, 0);
        }

        // take some time to release pressure before shutting of the motion app
        std::this_thread::sleep_for(std::chrono::milliseconds(int(exhaust_duration * 1000)));

        set_all_motion_apps(des_valve_mode, des_app_control);

        int actual_valve_mode, valve_state;

        int i = 0;
        // wait to reach valve mode 3 (pressure regulation) and valve state 2 (e.g. running)
        bool config_mismatch = true;
        while (config_mismatch) {
            if (i / sampling_rate > timeout) {
                return false;
            }

            config_mismatch = false;
            for (auto slot_idx = 0; slot_idx < (num_slots); slot_idx++) {
                get_single_motion_app(slot_idx, actual_valve_mode, valve_state);

                // std::cout << "Deactivate pressure reguluation idx: " << i << " for slot: " << slot_idx << " actual_valve_mode: " << actual_valve_mode << " valve_state: " << valve_state << std::endl;
                
                if (actual_valve_mode != des_valve_mode || valve_state != des_valve_state) {
                    config_mismatch = true;
                    break;
                }
            }

            if (config_mismatch) {
                i += 1;
                std::this_thread::sleep_for(std::chrono::milliseconds(int(1/sampling_rate*1000)));
            }
        }
        return true;
    } else {
        // Set valves to 0 bar (off).
        set_single_pressure(2*slot_idx, 0);
        set_single_pressure(2*slot_idx + 1, 0);

        // take some time to release pressure before shutting of the motion app
        std::this_thread::sleep_for(std::chrono::milliseconds(int(exhaust_duration * 1000)));

        set_single_motion_app(slot_idx, des_valve_mode, des_app_control);

        int actual_valve_mode, valve_state;
        get_single_motion_app(slot_idx, actual_valve_mode, valve_state);

        int i = 0;
        while (actual_valve_mode != des_valve_mode || valve_state != des_valve_state) {
            if (i / sampling_rate > timeout) {
                return false;
            }

            get_single_motion_app(slot_idx, actual_valve_mode, valve_state);

            // std::cout << "Deactivate pressure reguluation idx: " << i << " actual_valve_mode: " << actual_valve_mode << " valve_state: " << valve_state << std::endl;

            i += 1;
            std::this_thread::sleep_for(std::chrono::milliseconds(int(1/sampling_rate*1000)));
        }
        return true;
    }
}

void vtem_control::VtemControl::ensure_connection() const {
    if (!connected_) {
        throw std::runtime_error("Operation requires a connection.");
    }
}

void vtem_control::VtemControl::ensure_motion_app(int slot_idx, int des_motion_app_id, int des_valve_state) {
    int motion_app_id, valve_state;

    get_single_motion_app(slot_idx, motion_app_id, valve_state);

    if (motion_app_id != des_motion_app_id) {
        throw std::runtime_error("Operation requires activating the suitable motion app");
    }

    if (valve_state != des_valve_state) {
        throw std::runtime_error("Operation requires setting the desired valve state");
    }
}

int vtem_control::VtemControl::get_slot_idx_from_valve_idx(const int valve_idx) const {
    int slot_idx = std::floor(valve_idx / 2); // index of slot [0-7]
    return slot_idx;
}

int vtem_control::VtemControl::get_single_pressure(const int valve_idx) {
    ensure_connection();

    int slot_idx = get_slot_idx_from_valve_idx(valve_idx);
    int slot_remain = valve_idx - 2*slot_idx; // either 0 or 1 for valve in slot
    ensure_motion_app(slot_idx, 3, 2); // motion app pressure regulation with valve state "running"

    const auto dest = &input_value_buffer_[valve_idx];
    const auto addr = address_input_start + cpx_input_offset + 3*slot_idx + 1 + 1*slot_remain;

    if (modbus_read_registers(ctx_, addr, 1, (uint16_t *)dest) == -1) {
        throw std::runtime_error("Failed to read CPX modbus register.");
    }

    return *dest;
}

void vtem_control::VtemControl::set_single_pressure(const int valve_idx, const int pressure = 0) {
    ensure_connection();

    int slot_idx = get_slot_idx_from_valve_idx(valve_idx);
    int slot_remain = valve_idx - 2*slot_idx; // either 0 or 1 for valve in slot
    ensure_motion_app(slot_idx, 3, 2); // motion app pressure regulation with valve state "running"

    const auto addr = address_output_start + cpx_output_offset + 3*slot_idx + 1 + slot_remain;

    if (modbus_write_register(ctx_, addr, pressure) == -1) {
        throw std::runtime_error("Failed to write CPX modbus register.");
    }
}

void vtem_control::VtemControl::get_all_pressures(std::vector<int> *output) {
    ensure_connection();

    for (auto valve_idx = 0; valve_idx < 2*num_slots; valve_idx++) {
        get_single_pressure(valve_idx);
        output->at(valve_idx) = input_value_buffer_[valve_idx];
    }
}

void vtem_control::VtemControl::set_all_pressures(const std::vector<int> &pressures) {
    ensure_connection();

    for (auto valve_idx = 0; valve_idx < 2*num_slots; valve_idx++) {
        set_single_pressure(valve_idx, pressures[valve_idx]);
    }
}